src/pubsub.ts

```diff
@@ -57,56 +59,65 @@ export class AMQPPubSub implements PubSubEngine {
     if (refs && refs.length > 0) {
       const newRefs = [...refs, id];
       this.subsRefsMap[routingKey] = newRefs;
-      return Promise.resolve(id);
-    } else {
-      return this.subscriber.subscribe(this.exchange, routingKey, this.onMessage.bind(this))
-      .then(disposer => {
-        this.subsRefsMap[routingKey] = [
-          ...(this.subsRefsMap[routingKey] || []),
-          id,
-        ];
-        if (this.unsubscribeMap[routingKey]) {
-          return disposer();
-        }
-        this.unsubscribeMap[routingKey] = disposer;
-        return Promise.resolve(id);
-      });
+      return id;
     }
+
+    this.subsRefsMap[routingKey] = [
+      ...(this.subsRefsMap[routingKey] || []),
+      id,
+    ];
+
+    const existingDispose = this.unsubscribeMap[routingKey];
+    // Get rid of exisiting subscription while we get a new one.
+    const [newDispose] = await Promise.all([
+      this.subscriber.subscribe(this.exchange, routingKey, this.onMessage),
+      existingDispose ? existingDispose() : Promise.resolve(),
+    ]);
+
+    this.unsubscribeMap[routingKey] = newDispose;
+    return id;
   }
 
-  public unsubscribe(subId: number): Promise<void> {
-    const routingKey = this.subscriptionMap[subId].routingKey;
-    const refs = this.subsRefsMap[routingKey];
+  public async unsubscribe(subId: number): Promise<void> {
+    logger('Unsubscribe id: "%s"', subId);
```

Comment:
What's the reasoning behind this log line?

Reply:
See https://github.com/Surnet/graphql-amqp-subscriptions/pull/13/files#r423216743